docs/proposals/201909_thanos_sharding.md

```diff
@@ -0,0 +1,178 @@
+    * Requires separate design.
+
+This design is orthogonal for multi-buckets/multi-prefix support. 
+
+## Proposal
+
+We identified two use cases for relabeling on Thanos components:
+
+* Sharding: Operating on portion of object storage.
+* Advertising: Advertising custom labels on Info. This will result in pre-filtering on Querier which will not touch the server
+
+### Sharding
+
+On each component that works on the object storage (e.g Store GW and Compactor), add `--selector.relabel-config` (and corresponding `--selector.relabel-config-file`) that will
+be used to filter out what blocks should be selected for operations. Examples:
+
+* We want to run Compactor only for blocks with `external_labels` being `cluster=A`. We will run second Compactor for blocks with `cluster=B` external labels.
+* We want to browse only blocks with `external_labels` being `cluster=A` from object storage. We will run StoreGateway with selector of `cluster=A` from external labels of blocks.
+
+### Advertising
+
+On each component that exposes StoreAPI (e.g Querier, Ruler, Receiver, Store GW), add `--selector.relabel-config` (and corresponding `--selector.relabel-config-file`) that will
+be used to filter out what blocks should be selected for operations. Examples:
+
+* We want to run Ruler that will produce blocks with `external_labels` being `ruler_cluster="A", ruler_replica="1"`. This will ensure only those 
+
+NOTE: `selector.*` means that for those two use cases (Adv vs Sharding) we will use the same flags. This makes sense as essentially those are all about `selecting` what subset of metrics is given component responsible for.
+
+### Relabelling
+
+Similar to [promtail](https://github.com/grafana/loki/blob/master/docs/promtail.md#scrape-configs) this config will follow native
+[Prometheus relabel-config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config) syntax.
+
+The relabel config will define filtering process done on **every** synchronization with object storage.
+
+We will allow manipulating with several of inputs:
+
+* `__block_id`
+* External labels:
+  * `<name>` 
+* `__block_objstore_bucket_endpoint` 
+* `__block_objstore_bucket_name`
+* `__block_objstore_bucket_path`
+
+Output:
+
+* External labels to be advertised for this block.
+* If output is empty, drop block. 
+
+By default, on empty relabel-config, all external labels are assumed.
+Intuitively blocks without any external labels will be ignored.
+
+Example usages would be:
+
+* Drop blocks which contains external labels cluster=A
+```yaml
+- action: drop
+  regex: "A"
+  source_labels:
+  - cluster
+```
+* Keep only blocks which contains external labels cluster=A
+```yaml
+- action: keep
+  regex: "A"
+  source_labels:
+  - cluster
+```
+* Drop cluster label from external labels for each blocks (if present).
+```yaml
+- action: labeldrop
+  source_labels:
+  - cluster
```

Comment:
Action `labeldrop` should use `regex` instead.
```
- action: labeldrop
   regex: "^cluster$"
```

Reply:
good point!

Reply:
But also this is for next step. Updating this doc now (: